### PR TITLE
[security] Unsafe Inline Script

### PR DESCRIPTION
**Pull request by tabraiz12**
Have Closed the existing PR, splitting the PR to smaller PR for easier Review:

Upcoming PRs:
1. Unsafe Inline Script
2. Unsafe Inline Style
3. Backend Configuration for per page unsafe-eval, this is so that we will initially only allow editor.
4. Unsafe Eval - Editor Page, Login
5. Unsafe Eval - Admin Page
6. Unsafe-Eval - NoteBook, and others

## What changes were proposed in this pull request?

- Added a middleware to add request.nonce and include it in the request script
- Added unsafe-eval in CSP for temporarily handling the ko bindings, this fix will be part of another PR.

## How was this patch tested?

1. Manual Test
2. Testes Login
7. Tested Editor and running few commands
8. Navigated to job browser and admin pages didn't find any js issues.
9. Removed the nonce attribute and verified that its throwing error without those.

Please review [Hue Contributing Guide](https://github.com/cloudera/hue/blob/master/CONTRIBUTING.md) before opening a pull request.

### CODE REVIEW

File: desktop/conf.dist/hue.ini [hunk 1]
```diff
@@ -137,6 +137,9 @@ http_500_debug_mode=false
 # X-Content-Type-Options: nosniff This is a HTTP response header feature that helps prevent attacks based on MIME-type confusion.
 ## secure_content_security_policy="script-src 'self' 'unsafe-inline' 'unsafe-eval' *.googletagmanager.com *.doubleclick.net data:;img-src 'self' *.doubleclick.net http://*.tile.osm.org *.tile.osm.org *.gstatic.com data:;style-src 'self' 'unsafe-inline' fonts.googleapis.com;connect-src 'self' *.google-analytics.com;frame-src *;child-src 'self' data: *.vimeo.com;object-src 'none'"
 
+# Enable nonce attribute to remove unsafe-inline and auto remove unsafe-inline from csp 
+## csp_nonce=True
```
Comment (Harshg999) [inline desktop/conf.dist/hue.ini:141]: Nit: `true`

File: desktop/conf/pseudo-distributed.ini.tmpl [hunk 2]
```diff
@@ -142,6 +142,10 @@
   # X-Content-Type-Options: nosniff This is a HTTP response header feature that helps prevent attacks based on MIME-type confusion.
   ## secure_content_security_policy="script-src 'self' 'unsafe-inline' 'unsafe-eval' *.googletagmanager.com *.doubleclick.net data:;img-src 'self' *.doubleclick.net http://*.tile.osm.org *.tile.osm.org *.gstatic.com data:;style-src 'self' 'unsafe-inline' fonts.googleapis.com;connect-src 'self' *.google-analytics.com;frame-src *;child-src 'self' data: *.vimeo.com;object-src 'none'"
 
+  # Enable nonce attribute to remove unsafe-inline and auto remove unsafe-inline from csp 
+  ## csp_nonce=True
```
Comment (Harshg999) [inline desktop/conf/pseudo-distributed.ini.tmpl:146]: Nit: `true`

File: desktop/core/src/desktop/lib/django_util.py [hunk 3]
```diff
@@ -518,3 +518,12 @@ def __init__(self, data, encoder=DjangoJSONEncoder, safe=True,
     kwargs.setdefault('content_type', 'application/json')
     data = json.dumps(data, cls=encoder, **json_dumps_params)
     super(JsonResponse, self).__init__(content=data, **kwargs)
+
+
+def nonce_attribute(request):
+    # Assuming 'csp_nonce' is set in the request and CSP_NONCE controls its usage
+    if desktop.conf.CSP_NONCE.get():
```
Comment (Harshg999) [inline desktop/core/src/desktop/lib/django_util.py:525]: Instead of assuming, why not check for it by using `hasattr()`?

File: desktop/core/src/desktop/lib/django_util.py [hunk 4]
```diff
@@ -518,3 +518,12 @@ def __init__(self, data, encoder=DjangoJSONEncoder, safe=True,
...
+        csp_nonce = request.csp_nonce
+        if csp_nonce:
+            return f' nonce={csp_nonce}'
+    return ''
```
Comment (Harshg999) [inline desktop/core/src/desktop/lib/django_util.py:529]: Nit: Try something like this once?
```python
return f' nonce={csp_nonce}' if csp_nonce else ''
```

File: desktop/core/src/desktop/middleware.py [hunk 5]
```diff
@@ -798,18 +864,60 @@ def process_response(self, request, response):
...
+    def __init__(self, get_response):
+        self.get_response = get_response
+        self.secure_content_security_policy = SECURE_CONTENT_SECURITY_POLICY.get()
+        if not self.secure_content_security_policy:
+            LOG.info('Unloading ContentSecurityPolicyMiddleware')
+            raise exceptions.MiddlewareNotUsed
+
+    def process_request(self, request):
+        nonce = secrets.token_urlsafe()
+        request.csp_nonce = nonce
```
Comment (Harshg999) [inline desktop/core/src/desktop/middleware.py:876]: Nice!

**Review by Harshg999 — COMMENTED** [review]
Looking good! I've added few review comments..

File: desktop/core/src/desktop/js/onePageViewModel.js [hunk 6]
```diff
@@ -182,14 +182,12 @@ class OnePageViewModel {
       $.ajax({
         url: scriptUrl,
         converters: {
-          'text script': function (text) {
-            return text;
-          }
+          'text script': text => text
```
Comment (bjornalm) [inline desktop/core/src/desktop/js/onePageViewModel.js:185]: Just out of curiosity, was this change needed or just an improvement?

**Review by bjornalm — APPROVED** [review]
LGTM

File: desktop/core/src/desktop/templates/hue.mako [hunk 7]
```diff
@@ -295,8 +296,9 @@ ${ hueIcons.symbols() }
 </div>
 ${ commonshare() | n,unicode }
 
-% for bundle in get_hue_bundles('hue'):
-  ${ render_bundle(bundle) | n,unicode }
+% for bundle in get_hue_bundles('login' if section == 'login' else 'hue', 'LOGIN' if section == 'login' else 'DEFAULT'):
```
Comment (JohanAhlen) [inline desktop/core/src/desktop/templates/hue.mako:299]: Why the login related checks for hue.mako?

File: desktop/libs/notebook/src/notebook/templates/editor_components.mako [hunk 8]
```diff
@@ -2052,215 +2053,11 @@ else:
 
 
 <%def name="commonJS(is_embeddable=False, bindableElement='editorComponents', suffix='')">
+  <script type="application/json" id="editorOptionsJson">
+    ${ options_json | n,unicode,antixss }
+  </script>
+  <script ${nonce_attribute(request)} src="${ static('desktop/js/editor-component.js') }"></script>
```
Comment (JohanAhlen) [inline desktop/libs/notebook/src/notebook/templates/editor_components.mako:2059]: Missing js file in the commit

File: desktop/core/src/desktop/templates/hue.mako (hunk 7 above)
Comment (tabraiz12) [reply desktop/core/src/desktop/templates/hue.mako]: reverted

File: desktop/libs/notebook/src/notebook/templates/editor_components.mako (hunk 8 above)
Comment (tabraiz12) [reply desktop/libs/notebook/src/notebook/templates/editor_components.mako:2059]: added

File: desktop/core/src/desktop/js/onePageViewModel.js (hunk 6 above)
Comment (tabraiz12) [reply desktop/core/src/desktop/js/onePageViewModel.js:185]: this was not necessary, just an improvement.

**Review by JohanAhlen — APPROVED** [review]
LGTM! Just a question regarding process_response..

File: desktop/core/src/desktop/middleware.py [hunk 9]
```diff
@@ -798,18 +864,60 @@ def process_response(self, request, response):
...
+                new_csp.append(p)
 
-  def process_response(self, request, response):
-    if self.secure_content_security_policy and 'Content-Security-Policy' not in response:
-      response["Content-Security-Policy"] = self.secure_content_security_policy
+        # Update the Content-Security-Policy header with the new CSP string
+        response[header['name']] = "; ".join(new_csp).strip() + ';'
 
-    return response
+        return response
```
[the remark this answers is not in the document]
Comment (JohanAhlen) [reply desktop/core/src/desktop/middleware.py:920]: But with this indentation change it'll only return the response if the condition is true right?

[the remark this answers is not in the document]
Comment (tabraiz12) [reply desktop/core/src/desktop/middleware.py]: i had used 4 space to indent instead of 2 space as in file, hence there is a diff, i have changed everything back to 2 spaces.
